File: node/src/benchmarks.js
const yargs = require("yargs");
const fs = require("fs");
const path = require("path");
const Benchmark = require("./lib/benchmark");
// Functions
const concat = require("./functions/concat");
const counter = require("./functions/counter");
const arrayfill = require("./functions/arrayfill");
const mapfill = require("./functions/mapfill");
const jsondecode = require("./functions/jsondecode");
const fileread = require("./functions/fileread");

let jsonData = {};

// Get iterations
const argv = yargs.option("iter", {
  alias: "i",
  description: "set number of iterations for the benchmark",
  type: Number,
}).argv;

let iterations = argv.iter;

// Guard against zero/negative iterations
if (iterations < 1) {
  iterations = 100;
}

// Setup
try {
  jsonData = fs.readFileSync(path.resolve(__dirname, "./data/demo.json"));
} catch (err) {
  console.error("Error reading JSON file:", err);
}

// Benchmark
let benchmark = new Benchmark();
benchmark
  .setIter(iterations)
  .add("Concatenate Strings", concat)
  .add("Counter", counter)
  .add("Array Fill", arrayfill)
  .add("Map Fill", mapfill)
  .add("JSON Parse", jsondecode, jsonData)
  .add("File Read", fileread, path.resolve(__dirname, "./data/demo.txt"))
  .run();

// Write results to file
let resultsJSON = benchmark.getJSON();
try {
  fs.writeFileSync(`/results/results-node-cli-${iterations}.json`, resultsJSON);
} catch (err) {
  console.error("Error while writing results:", err);
}

File: go/src/benchmarks.go
package main

import (
	"benchmark-project/functions"
	"benchmark-project/lib"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
)

func main() {
	// Get Iterations
	var iterations int
	flag.IntVar(&iterations, "iter", 100, "Iterations to benchmark")
	flag.Parse()
	if iterations < 1 {
		iterations = 100
	}

	// Setup
	jsonData, err := ioutil.ReadFile("./data/demo.json")
	if err != nil {
		log.Fatalf("Error reading JSON file: %s", err.Error())
	}

	// Benchmark
	benchmark := lib.Benchmark{}
	benchmark.SetIter(iterations)
	benchmark.Add("Concatenate Strings", functions.Concat)
	benchmark.Add("Counter", functions.Counter)
	benchmark.Add("Array Fill", functions.Arrayfill)
	benchmark.Add("Map Fill", functions.Mapfill)
	benchmark.Add("JSON Parse", functions.Jsondecode, string(jsonData))
	benchmark.Add("File Read", functions.Fileread, "./data/demo.txt")
	benchmark.Run()

	// Write results to file
	resultsJSON := benchmark.GetJSON()
	f, err := os.OpenFile(fmt.Sprintf("/results/results-go-cli-%s.json", fmt.Sprint(iterations)), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
	if err != nil {
		log.Fatalf("Error opening JSON file for results: %s", err.Error())
	}
	_, err = f.Write(resultsJSON)
	if err != nil {
		log.Fatalf("Error writing JSON results file: %s", err.Error())
	}
}
